### PR TITLE
Make ReportingTestSet behaviour consistent

### LINKED ISSUE #34: Move display_reporting_testset inside finish of ReportingTestSet
This will allow the displayed output to match `Pkg.test` more closely, as it will be displayed before flattening occurs. 

Need to see what affect this will have if something like

```Julia
@testset ReportingTestSet begin @test 1==1 end
```

Is done however, as `display_reporting_testset` may assume a blank top level.

### PR DESCRIPTION
**Pull request by mmiller-max**
@oxinabox please let me know your thoughts on whether this is the right thing to do, I'm in two minds.

I started out moving `display_reporting_testset` to inside of `finish` (closes #34) and realised that both `display_reporting_testset` and `flatten_results` assume that they are being used by something like `TestReports.test`, i.e. that there is an additional `ReportingTestSet` around the tests that are being tested. (As does `report`). This wouldn't be the case if someone was using  a `ReportingTestSet` outside of this (which we don't recommend).

There are two options to make this consistent:

1. Don't export `ReportingTestSet` so that it is private to `TestReports`, which in turn means that `report`,  `display_reporting_testset` and `flatten_results` (the latter two by `finish`) are only used within `TestReports` and therefore doesn't matter what they expect.

2. Do what I've done here, which makes `TestReports.test` add a property to the wrapping `TestSet` which is then checked in `finish`. If the property is not there, an additional wrapping testset is added so that the flattened `ReportingTestSet` will work with `report` in the same way, and the tests will be displayed correctly.

2 gives more flexibility, but I'm not sure if its overkill and opening a can of worms by suggesting people could use `ReportingTestSets` independently. I do, however, like that it will structure any `ReportingTestSet` to be in the correct format for the JUnit XML, and display the results as per `Pkg.test`.

Either way `display_reporting_testset` should be moved inside of `finish` to improve the printing of results.

### CODE REVIEW

**Comment by oxinabox** [comment]
I think I am missing something about why this is needed to be recorded in a property.

Can't we use `get_testset` to determine if this is top level or not, and if not what that top level is?
https://github.com/JuliaTesting/TestReports.jl/blob/babb28f958e5152322e0476394a00690830eea53/src/testsets.jl#L15-L20
Couldn't we fix this in `display_reporting_testset` and `flatten_results` and `report` ?

**Comment by mmiller-max** [comment]
I'll try to take a step back!

Because the runner wraps `runtests.jl` in a `ReportingTestSet`, `flatten_results` and `display_reporting_testset` assume that this top level test set is just a wrapper and flatten the results of that testset/don't display it respectively. Then the output of the finished `ReportingTestSet` is three levels deep, `TestSet` -> `TestSet` -> `Results` which become Testsuites -> Testsuite -> Testcases respectively in `report`.

If a `ReportingTestSet` is used directly (which isn't recommended but it is exported), then the top level test set won't be a wrapper. 

In the case of

```Julia
ts = @testset ReportingTestSet "MyTopLevel" begin
    @test 1==1
    @test 2==2
end
```
you end up with

```
Julia> display(ts)
ts = ReportingTestSet("MyTopLevel", 
    Test.AbstractTestSet[
        ReportingTestSet("Top level tests", 
            Test.Result[Test Passed, Test Passed], 
            Dict{String,Any}()
        )
    ],
     Dict{String,Any}()
)
julia> display_reporting_testset(ts)
Test Summary:   | Pass  Total
Top level tests |    2      2
```
Once the test set is finished.

For the above situation to print/flatten/report correctly, I think these functions need to know whether or not there has been a wrapper, which is what I was suggesting that we use the property for. The thing the functions need to know is whether the top level test set is a wrapper, rather than whether or not a test set is top level.

As all three of these functions would need changing, I thought it was easier to change `finish` to add in this wrapper if it didn't exist. It also makes a `ReportingTestSet` always in the right structure for a JUnit XML.

Maybe a nicer way is to add a new field to `ReportingTestSet`s, something like `is_test_reports_wrapper` rather than using a property, or maybe we just don't export it and don't expect to handle `ReportingTestSet`s being used outside of the runner.

Hope that makes more sense!

**Comment by mmiller-max** [comment]
Yep that makes sense with the helper function. Would we say that "normal" is having the additional wrapping test set?

And no worries! I was doing it in `TestReports.test` as I'm not sure how `get_testset()` would be able to know whether the wrapper was being used or not - I think you could have a top level `ReportingTestSet` that had the same results/structure from both `TestReports.test` and when using it directly, and I couldn't think of another way to distinguish between the two.

e.g. With a `runtests.jl` file like this:
```Julia
@testset "mytests1" begin
    @test 1==1
end
@testset "mytests2" begin
    @test 1==1
end
```
`ts` in the wrapper would look the same as this code run directly:
```Julia
ts = @testset ReportingTestSet "" begin
    @testset "mytests1" begin
        @test 1==1
    end
    @testset "mytests2" begin
        @test 1==1
    end
end
```
But I could be missing something with `get_testset()`?

**Comment by oxinabox** [comment]
Ah right most tests i am writing these days use a [outer testset themselves.](https://github.com/invenia/NamedDims.jl/blob/be2d207346ca6b56a25cea31d65e2b257ac55d37/test/runtests.jl#L20)
Because otherwise when the first testset fails the others don't run,
and the default testset also messes up display when doing 1 at a time.


So we can examine the whole Testset stack using:
```
get_testset_stack() = get(task_local_storage(), :__BASETESTNEXT__, Test.AbstractTestSet[])
```
Which i thought was something we could do with `get_testset` but apprently that only peeks at the top of the stack.

Would it surfice to just check of the top of the stack is a `ReportingTestset` ?


```julia
julia> begin
       @testset "A" begin
       @show get_testset_stack()
       end
       @testset "B" begin
       @show get_testset_stack()
       end
       end;
get_testset_stack() = AbstractTestSet[Test.DefaultTestSet("A", Any[], 0, false)]
Test Summary: |
A             | No tests
get_testset_stack() = AbstractTestSet[Test.DefaultTestSet("B", Any[], 0, false)]
Test Summary: |
B             | No tests

julia> @testset "outer" begin
       @testset "A" begin
       @show get_testset_stack()
       end
       @testset "B" begin
       @show get_testset_stack()
       end
       end;
get_testset_stack() = AbstractTestSet[Test.DefaultTestSet("outer", Any[], 0, false), Test.DefaultTestSet("A", Any[], 0, false)]
get_testset_stack() = AbstractTestSet[Test.DefaultTestSet("outer", Any[Test.DefaultTestSet("A", Any[], 0, false)], 0, false), Test.DefaultTestSet("B", Any[], 0, false)]
Test Summary: |
outer         | No tests
```

**Comment by mmiller-max** [comment]
Sorry by "normal" I meant what are considering the normal way for a `ReportingTestSet` to be used, sorry I wasn't clear there.

I think first of all we need to decide if we want people to use `ReportingTestSet`s outside of `TestReports`. Based on the answer to this question, there are routes we could take:

### 1 We don't want people to use `ReportingTestSet`s outside of `TestReports`

I think we would then need to:

1. Stop exporting `ReportingTestSet`s so it's clear they're internal to the package
2. Put a check in `report` to make sure any `AbstractTestSet` given to it is compatible
3. We can leave `flatten_results` as we only need it to work for `TestReports`, which it does
4. Similarly we can leave `display_reporting_testset` (but I would move it inside `finish` as then it will print more closely to how `Pkg` prints)

### 2 We do want people to have the option of using `ReportingTestSet`s outside of `TestReports`

In this case we need to make sure a `ReportingTestSet` `finish`es correctly when used. Possible uses:

1. If one is used by `TestReports`, then for the purposes of displaying and flattening we want to ignore the top level `TestSet` as it's a wrapper added by `TestReports`.
2. If one is used in the default tests, for example inside `runtests.jl`, then we want to display the top level `TestSet` when `finish`ing. However I don't think anyone would want to use it in this way. Doesn't make much sense ...
3. I can only see someone wanting to use a `ReportingTestSet` directly if they are trying to generate their own report type, rather than a JUnit XML, in which case they would probably still wrap their `runtests.jl` in a `ReportingTestSet`, so that `Pkg.test` is unaffected and then they can use the wrapping `ReportingTestSet` for their own report generation. In fact we could add this to the docstring for a `ReportingTestSet`.

I don't think there's any point in allowing for 2, and 3 works in the same way as 1 so we will just need the docstring updating with suggested use of a `ReportingTestSet`.

### TL;DR

Sorry for another long post. Conclusion is: unless you have any objections, I propose the following:

- Add a note to the docstring of ReportingTestSet advising how it should be used if it is used directly (which will be to always wrap `runtests.jl` or similar)
- Move `display_reporting_testset` inside of `finish` so that a `ReportingTestSet` always displays and displays the results before flattening
- Add a check to `report` to make sure any input to it is compatible.

**Comment by oxinabox** [comment]
I agree with these conclusions

**Comment by mmiller-max** [comment]
Updated PR with changes as discussed

**Comment by mmiller-max** [comment]
@oxinabox please could I get a review? Cheers!

**Review by oxinabox — APPROVED** [review]
